tree/30_generic_methods/file_copy_from_local_source_with_check.cf

```diff
@@ -0,0 +1,70 @@
+#####################################################################################
+# Copyright 2018 Normation SAS
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name File copy from local source with check
+# @description Ensure that a file or directory is copied from a local source
+# if a check command succeeds
+#
+# @parameter source      Source file (absolute path on the target node)
+# @parameter destination Destination file (absolute path on the target node)
+# @parameter check_command   Command to run, it will get the source path as argument
+# @parameter rc_ok       Return codes to be considered as valid, separated by a comma (default is 0)
+# 
+# @class_prefix file_copy_from_local_source
+# @class_parameter destination
+
+bundle agent file_copy_from_local_source_with_check(source, destination, check_command, rc_ok)
+{
+  vars:
+      "old_class_prefix" string => canonify("file_copy_from_local_source_with_check_${destination}");
+      "promisers"          slist => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist => { "${source}", "${destination}", "${check_command}" };
+      "canonified_file"   string => canonify("${destination}");
+
+  defaults:
+      "rc_ok"             string => "0", if_match_regex => "";
+
+  classes:
+      "copy_needed" not => returnszero("diff \"${source}\" \"${destination}\"", "useshell");
+
+      "pass3"  expression => "pass2";
+      "pass2"  expression => "pass1";
+      "pass1"  expression => "any";
+
+  methods:
+      "check"  usebundle => condition_from_command("file_copy_from_local_source_with_check_${canonified_file}_check", "${check_command} \"${source}\"", "${rc_ok}", "999"),
```

Comment:
same here, we want the condition to be run only if the class agent is set (meaning, only if agent is running, not cf-promises, or any other ocmponent)
that would avoid triggering unwanted actions that might have side effet
(on a side note, it may be relevant to add this in the generic methods that runs actions also)

Reply:
IIRC `commands` promises are not executed by cf-promises, only `returnszero` and `execresult`

Reply:
you are totally right